### PR TITLE
KAFKA-5166: Add option "dry run" to Streams application reset tool

### PR DESCRIPTION
**Pull request by bharatviswa504**
Addressed the below review comment from #PR #2998 from @mjsax 

I am wondering if it would be better, to "embed" the dry-run into the actual code and branch on each place. Otherwise, if things get changed, we could easily introduce bugs (ie, dry run show something different than what the actual reset code does.

We could introduce methods like mabyeSeekToBeginning() that either does the seek or only prints to stdout. This would ensure that the main logic is used to "feed" into dry-run and we don't have code duplication.

WDYT?

### CODE REVIEW

**Comment by bharatviswa504** [comment]
@mjsax I initially thought it will be good. But I thought the code will not look cleaner it will have a lot of if else, for simple logic. But as you said, if main logic changes, if someone missed to update dryRun it will deviate from actual behavior. Update the code accordingly.

**Comment by mjsax** [comment]
I just realized that the JavaDoc of this class is not correct: Can you please fix this as a side task?

It says: "setting the application's consumer offsets for input and internal topics to zero". This should be "setting the application's consumer offsets for input topics to the beginning of the log". And we should also add to bullet point about "deleted topic" that the corresponding offsets will be deleted automatically.

File: core/src/main/scala/kafka/tools/StreamsResetter.java [hunk 1]
```diff
@@ -157,49 +164,63 @@ private void parseArguments(final String[] args) throws IOException {
...
+        Boolean dryRun = options.has(dryRunOption);
+        String groupId = options.valueOf(applicationIdOption);
+
+        if (!dryRun) {
+            if (inputTopics.size() == 0 && intermediateTopics.size() == 0) {
+                System.out.println("No input or intermediate topics specified. Skipping seek.");
+                return;
+            } else {
+                if (inputTopics.size() != 0) {
+                    System.out.println("Resetting offsets to zero for input topics " + inputTopics + " and all internal topics.");
```
Comment (mjsax) [inline core/src/main/scala/kafka/tools/StreamsResetter.java]: Can you please fix this output too -- the tool does "seek to beginning" and does not set offsets to zero.

File: core/src/main/scala/kafka/tools/StreamsResetter.java [hunk 2]
```diff
@@ -219,23 +240,85 @@ private void resetInputAndInternalAndSeekToEndIntermediateTopicOffsets() {
...
-            for (final TopicPartition p : partitions) {
-                client.position(p);
+            if (notFoundIntermediateTopics.size() > 0) {
+                System.out.println("Following intermediate topics are not found, skipping them");
+                for (String topic : notFoundIntermediateTopics) {
+                    System.out.println("Topic - " + topic);
+                }
             }
-            client.commitSync();
+            client.close();
```
Comment (mjsax) [inline core/src/main/scala/kafka/tools/StreamsResetter.java]: Not required. Client will be automatically closes, as we use Java's "try with resource" feature.

File: core/src/main/scala/kafka/tools/StreamsResetter.java [hunk 3]
```diff
@@ -246,24 +329,36 @@ private boolean isIntermediateTopic(final String topic) {
...
-                    "--delete", "--topic", topic});
                 try {
-                    TopicCommand.deleteTopic(zkUtils, commandOptions);
+                    if (!dryRun) {
+                        final TopicCommand.TopicCommandOptions commandOptions = new TopicCommand.TopicCommandOptions(new String[]{
+                            "--zookeeper", options.valueOf(zookeeperOption),
+                            "--delete", "--topic", topic});
+                        TopicCommand.deleteTopic(zkUtils, commandOptions);
+                    } else {
+                        System.out.println("Topic - " + topic);
```
Comment (mjsax) [inline core/src/main/scala/kafka/tools/StreamsResetter.java]: Nit: maybe (`"Topic: " + topic)` ?

File: core/src/main/scala/kafka/tools/StreamsResetter.java [hunk 4]
```diff
@@ -246,24 +329,36 @@ private boolean isIntermediateTopic(final String topic) {
...
+                    }
                 } catch (final RuntimeException e) {
                     System.err.println("ERROR: Deleting topic " + topic + " failed.");
                     throw e;
                 }
             }
         }
 
-        System.out.println("Done.");
+        if (!dryRun) {
```
Comment (mjsax) [inline core/src/main/scala/kafka/tools/StreamsResetter.java]: Do we need to suppress this in dry run?

File: core/src/main/scala/kafka/tools/StreamsResetter.java [hunk 5]
```diff
@@ -246,24 +329,36 @@ private boolean isIntermediateTopic(final String topic) {
         return options.valuesOf(intermediateTopicsOption).contains(topic);
     }
 
-    private void deleteInternalTopics(final ZkUtils zkUtils) {
-        System.out.println("Deleting all internal/auto-created topics for application " + options.valueOf(applicationIdOption));
+    private void maybedeleteInternalTopics(final ZkUtils zkUtils) {
+        Boolean dryRun = options.has(dryRunOption);
```
Comment (mjsax) [inline core/src/main/scala/kafka/tools/StreamsResetter.java]: nit: add `final`
Why not use `boolean` ?

File: core/src/main/scala/kafka/tools/StreamsResetter.java [hunk 6]
```diff
@@ -246,24 +329,36 @@ private boolean isIntermediateTopic(final String topic) {
...
+
+        if (!dryRun) {
+            System.out.println("Deleting all internal/auto-created topics for application " + options.valueOf(applicationIdOption));
+        } else {
+            System.out.println("Following internal/auto-created topics will be Deleted for application " + options.valueOf(applicationIdOption));
```
Comment (mjsax) [inline core/src/main/scala/kafka/tools/StreamsResetter.java]: I think this message is not much different from the non-dry-run one. Maybe we can have one "unified" message that can be used for both cases (to reduce the number of `if-else` branches?

File: core/src/main/scala/kafka/tools/StreamsResetter.java [hunk 7]
```diff
@@ -219,23 +240,85 @@ private void resetInputAndInternalAndSeekToEndIntermediateTopicOffsets() {
...
+                    if (allTopics.contains(topic)) {
+                        System.out.println("Topic - " + topic);
+                    }
+                }
+            }
+        }
+
+    }
+
+    private void maybeSeekToBeginning(KafkaConsumer<byte[], byte[]> client, Set<TopicPartition> inputAndInternalTopicPartitions, List<String> internalTopics) {
```
Comment (mjsax) [inline core/src/main/scala/kafka/tools/StreamsResetter.java]: nit: add `final` to all parameters -- also, the line is quite long, maybe start a new line for each parameter?

File: core/src/main/scala/kafka/tools/StreamsResetter.java [hunk 8]
```diff
@@ -219,23 +240,85 @@ private void resetInputAndInternalAndSeekToEndIntermediateTopicOffsets() {
...
+
+        final Boolean dryRun = options.has(dryRunOption);
+        final List<String> inputTopics = options.valuesOf(inputTopicsOption);
+        final String groupId = options.valueOf(applicationIdOption);
+
+        if (inputAndInternalTopicPartitions.size() > 0) {
+            if (!dryRun) {
+                client.seekToBeginning(inputAndInternalTopicPartitions);
+            } else {
+                System.out.println("Following input topics offsets will be reset to beginning (for consumer group " + groupId + ")");
```
Comment (mjsax) [inline core/src/main/scala/kafka/tools/StreamsResetter.java:312]: Nit: go we need to include the groupId? It's repeated in line 316.

File: core/src/main/scala/kafka/tools/StreamsResetter.java [hunk 9]
```diff
@@ -219,23 +240,85 @@ private void resetInputAndInternalAndSeekToEndIntermediateTopicOffsets() {
...
             throw e;
         }
 
-        System.out.println("Done.");
+        if (!dryRun) {
+            System.out.println("Done.");
+        }
+    }
+
+    private void maybeSeekToEnd(KafkaConsumer<byte[], byte[]> client, Set<TopicPartition> intermediateTopicPartitions) {
```
Comment (mjsax) [inline core/src/main/scala/kafka/tools/StreamsResetter.java]: Nit: add `final` to parameters

File: core/src/main/scala/kafka/tools/StreamsResetter.java [hunk 10]
```diff
@@ -219,23 +240,85 @@ private void resetInputAndInternalAndSeekToEndIntermediateTopicOffsets() {
...
+
+        final Boolean dryRun = options.has(dryRunOption);
```
Comment (mjsax) [inline core/src/main/scala/kafka/tools/StreamsResetter.java]: Nit: why not use `boolean` ?

File: core/src/main/scala/kafka/tools/StreamsResetter.java [hunk 11]
```diff
@@ -219,23 +240,85 @@ private void resetInputAndInternalAndSeekToEndIntermediateTopicOffsets() {
...
+                    }
+                }
+            }
+        }
+
+    }
+
+    private void maybeSeekToBeginning(KafkaConsumer<byte[], byte[]> client, Set<TopicPartition> inputAndInternalTopicPartitions, List<String> internalTopics) {
+
+        final Boolean dryRun = options.has(dryRunOption);
```
Comment (mjsax) [inline core/src/main/scala/kafka/tools/StreamsResetter.java]: Nit: why not use boolean

File: core/src/main/scala/kafka/tools/StreamsResetter.java [hunk 12]
```diff
@@ -219,23 +240,85 @@ private void resetInputAndInternalAndSeekToEndIntermediateTopicOffsets() {
...
+        final String groupId = options.valueOf(applicationIdOption);
+        final List<String> intermediateTopics = options.valuesOf(intermediateTopicsOption);
+
+        if (intermediateTopicPartitions.size() > 0) {
+            if (!dryRun) {
+                client.seekToEnd(intermediateTopicPartitions);
+            } else {
+                System.out.println("Following intermediate topics offsets will be reset to end (for consumer group " + groupId + ")");
+                for (String topic : intermediateTopics) {
```
Comment (mjsax) [inline core/src/main/scala/kafka/tools/StreamsResetter.java]: nit: add `final`

File: core/src/main/scala/kafka/tools/StreamsResetter.java [hunk 13]
```diff
@@ -104,8 +103,15 @@ public int run(final String[] args, final Properties config) {
...
-            resetInputAndInternalAndSeekToEndIntermediateTopicOffsets();
-            deleteInternalTopics(zkUtils);
+
+            if (!adminClient.describeConsumerGroup(groupId, 0).consumers().get().isEmpty()) {
+                throw new IllegalStateException("Consumer group '" + groupId + "' is still active. " +
+                            "Make sure to stop all running application instances before running the reset tool.");
+            }
+
+            mayberesetInputAndInternalAndSeekToEndIntermediateTopicOffsets();
+            maybedeleteInternalTopics(zkUtils);
```
Comment (mjsax) [inline core/src/main/scala/kafka/tools/StreamsResetter.java]: nit: `maybeDeleteInternalTopics` and `maybeReset...`

File: core/src/main/scala/kafka/tools/StreamsResetter.java [hunk 14]
```diff
@@ -157,49 +164,63 @@ private void parseArguments(final String[] args) throws IOException {
...
-                System.out.println("Seek-to-end for intermediate topics " + intermediateTopics);
+        final List<String> internalTopics = new ArrayList<>();
+
+        final List<String> notFoundInputTopics = new ArrayList<>();
+        final List<String> notFoundIntermediateTopics = new ArrayList<>();
+
+        Boolean dryRun = options.has(dryRunOption);
+        String groupId = options.valueOf(applicationIdOption);
+
+        if (!dryRun) {
```
Comment (mjsax) [inline core/src/main/scala/kafka/tools/StreamsResetter.java]: Not sure why we need this here? If no input topics are specified, we can just return in dry-run case, too?

File: core/src/main/scala/kafka/tools/StreamsResetter.java [hunk 15]
```diff
@@ -219,23 +240,85 @@ private void resetInputAndInternalAndSeekToEndIntermediateTopicOffsets() {
...
+                }
+                client.commitSync();
             }
-            if (intermediateTopicPartitions.size() > 0) {
-                client.seekToEnd(intermediateTopicPartitions);
+
+            if (notFoundInputTopics.size() > 0) {
+                System.out.println("Following input topics are not found, skipping them");
+                for (String topic : notFoundInputTopics) {
+                    System.out.println("Topic - " + topic);
```
Comment (mjsax) [inline core/src/main/scala/kafka/tools/StreamsResetter.java]: Nit: maybe `("Topic: " + topic)`

File: core/src/main/scala/kafka/tools/StreamsResetter.java [hunk 16]
```diff
@@ -219,23 +240,85 @@ private void resetInputAndInternalAndSeekToEndIntermediateTopicOffsets() {
...
+                }
             }
 
-            for (final TopicPartition p : partitions) {
-                client.position(p);
+            if (notFoundIntermediateTopics.size() > 0) {
+                System.out.println("Following intermediate topics are not found, skipping them");
+                for (String topic : notFoundIntermediateTopics) {
+                    System.out.println("Topic - " + topic);
```
Comment (mjsax) [inline core/src/main/scala/kafka/tools/StreamsResetter.java]: Nit: maybe `("Topic: " + topic)`

File: core/src/main/scala/kafka/tools/StreamsResetter.java [hunk 17]
```diff
@@ -157,49 +164,63 @@ private void parseArguments(final String[] args) throws IOException {
...
-        config.setProperty(ConsumerConfig.BOOTSTRAP_SERVERS_CONFIG, options.valueOf(bootstrapServerOption));
-        config.setProperty(ConsumerConfig.GROUP_ID_CONFIG, options.valueOf(applicationIdOption));
-        config.setProperty(ConsumerConfig.ENABLE_AUTO_COMMIT_CONFIG, "false");
-
         final Set<String> topicsToSubscribe = new HashSet<>(inputTopics.size() + intermediateTopics.size());
+
         for (final String topic : inputTopics) {
             if (!allTopics.contains(topic)) {
-                System.err.println("Input topic " + topic + " not found. Skipping.");
+                notFoundInputTopics.add(topic);
```
Comment (mjsax) [inline core/src/main/scala/kafka/tools/StreamsResetter.java:205]: Why this change?

File: core/src/main/scala/kafka/tools/StreamsResetter.java [hunk 18]
```diff
@@ -157,49 +164,63 @@ private void parseArguments(final String[] args) throws IOException {
...
             } else {
                 topicsToSubscribe.add(topic);
             }
         }
         for (final String topic : intermediateTopics) {
             if (!allTopics.contains(topic)) {
-                System.err.println("Intermediate topic " + topic + " not found. Skipping.");
+                notFoundIntermediateTopics.add(topic);
```
Comment (mjsax) [inline core/src/main/scala/kafka/tools/StreamsResetter.java:212]: Why this change?

File: core/src/main/scala/kafka/tools/StreamsResetter.java [hunk 19]
```diff
@@ -157,49 +164,63 @@ private void parseArguments(final String[] args) throws IOException {
...
-                System.out.println("Resetting offsets to zero for input topics " + inputTopics + " and all internal topics.");
-            }
-            if (intermediateTopics.size() != 0) {
-                System.out.println("Seek-to-end for intermediate topics " + intermediateTopics);
+        final List<String> internalTopics = new ArrayList<>();
+
+        final List<String> notFoundInputTopics = new ArrayList<>();
+        final List<String> notFoundIntermediateTopics = new ArrayList<>();
+
+        Boolean dryRun = options.has(dryRunOption);
```
Comment (mjsax) [inline core/src/main/scala/kafka/tools/StreamsResetter.java]: Nit: why not use `boolean`

**Review by mjsax — CHANGES_REQUESTED** [review]
Thanks for the update. Most comments are minor.

One more meta comment: might be better tp use one member variable `boolean dryRun` that is set after `parseArgument()` (L91) instead of all the local `dryRun` vars.

File: core/src/main/scala/kafka/tools/StreamsResetter.java (hunk 17 above)
Comment (bharatviswa504) [reply core/src/main/scala/kafka/tools/StreamsResetter.java:205]: This is done, to make output clear 
System.out.println("----Dry run displays the actions which will be performed when running Streams Reset Tool----");

As I am printing the above one, and then below printing all dry-run actions, so want to make it clear with description done like this.
Following input topics are not found, skipping them
Topic: t1
Following input topics are not found, skipping them
Topic:t2

File: core/src/main/scala/kafka/tools/StreamsResetter.java (hunk 8 above)
Comment (bharatviswa504) [reply core/src/main/scala/kafka/tools/StreamsResetter.java:312]: The line is added again to make clear distinguish
Following input topics offsets will be reset to beginning (for consumer group g1)
t1
Following internal topics offsets will be reset to beginning (for consumer group g1
t5

And also consumer group is shown, to provide info to user for this topic offsets are reset for this consumer group.
As most is done to make the output clear

File: core/src/main/scala/kafka/tools/StreamsResetter.java (hunk 6 above)
Comment (bharatviswa504) [reply core/src/main/scala/kafka/tools/StreamsResetter.java]: Added the check because, in dry-run we only print actions, and not do the action.
As this statement looks like Deleting all internal/auto-created topics for application g1 is like it is an action.

File: core/src/main/scala/kafka/tools/StreamsResetter.java (hunk 4 above)
Comment (bharatviswa504) [reply core/src/main/scala/kafka/tools/StreamsResetter.java]: Just added to make the output look clear.

As if it is not that major, removed the check to print done.

File: core/src/main/scala/kafka/tools/StreamsResetter.java (hunk 1 above)
Comment (bharatviswa504) [reply core/src/main/scala/kafka/tools/StreamsResetter.java]: fixed.

File: core/src/main/scala/kafka/tools/StreamsResetter.java (hunk 13 above)
Comment (bharatviswa504) [reply core/src/main/scala/kafka/tools/StreamsResetter.java]: changed.

File: core/src/main/scala/kafka/tools/StreamsResetter.java (hunk 19 above)
Comment (bharatviswa504) [reply core/src/main/scala/kafka/tools/StreamsResetter.java]: it can be. Changed it as suggested.

File: core/src/main/scala/kafka/tools/StreamsResetter.java (hunk 14 above)
Comment (bharatviswa504) [reply core/src/main/scala/kafka/tools/StreamsResetter.java]: Added the check to make output clear.
As with out checking input topics, existed or not, we print these input topics offset will be seek to end or beginning.
In dry-run i have taken care that only for topics, which exist print seek to end/begininng. for remaining topics print no action on them.

**Comment by bharatviswa504** [comment]
@mjsax Addressed most of the review comments.
Some changes are still left, as they are done to make dry-run output look cleaner.
Below is the dry run output
bin/kafka-streams-application-reset.sh --input-topics streams-file-input,t1,t9 --zookeeper localhost:2181 --dry-run

----Dry run displays the actions which will be performed when running Streams Reset Tool----
Following input topics offsets will be reset to beginning (for consumer group streams-wordcount)
Topic: streams-file-input
Following internal topics offsets will be reset to beginning (for consumer group streams-wordcount)
Topic: streams-wordcount-Counts-repartition
Topic: streams-wordcount-Counts-changelog
Following input topics are not found, skipping them
Topic: t1
Topic: t9
Done.
Following internal/auto-created topics will be Deleted for application streams-wordcount
Topic: streams-wordcount-Counts-repartition
Topic: streams-wordcount-Counts-changelog
Done.

And regarding cleanup for resetting offsets for internal topics, will be done as part of this JIRA 	KAFKA-5210. Once this PR is delivered in.
Please let me know any more changes needed.

File: core/src/main/scala/kafka/tools/StreamsResetter.java (hunk 18 above)
Comment (bharatviswa504) [reply core/src/main/scala/kafka/tools/StreamsResetter.java:212]: same as above comment.

**Comment by bharatviswa504** [comment]
retest this please

**Comment by bharatviswa504** [comment]
The test case failures are not related to these changes.

File: core/src/main/scala/kafka/tools/StreamsResetter.java (hunk 8 above)
Comment (mjsax) [inline core/src/main/scala/kafka/tools/StreamsResetter.java]: I understand the motivation and agree it's a good idea. It's only that 

`"Deleting all internal/auto-created topics for application <app.id>."`
and
`"Following internal/auto-created topics will be Deleted for application <app.id>."`

sound super similar. So not sure if we gain much. But it's quite subjective. It's ok if you leave as is. Please fix the typo though `Deleted` -> `deleted`.

File: core/src/main/scala/kafka/tools/StreamsResetter.java (hunk 6 above)
Comment (mjsax) [reply core/src/main/scala/kafka/tools/StreamsResetter.java]: I understand the motivation and agree it's a good idea. It's only that 

`"Deleting all internal/auto-created topics for application <app.id>."`
and
`"Following internal/auto-created topics will be Deleted for application <app.id>."`

sound super similar. So not sure if we gain much. But it's quite subjective. It's ok if you leave as is. Please fix the typo though `Deleted` -> `deleted`.

**Comment by mjsax** [comment]
Call for review @enothereska @dguy

File: core/src/main/scala/kafka/tools/StreamsResetter.java (hunk 6 above)
Comment (bharatviswa504) [reply core/src/main/scala/kafka/tools/StreamsResetter.java]: Agreed. Deleted the condition, as anyway there is this displayed on start, when run with dry-run option.
----Dry run displays the actions which will be performed when running Streams Reset Tool----

File: core/src/main/scala/kafka/tools/StreamsResetter.java [hunk 20]
```diff
@@ -16,10 +16,18 @@
  */
 package kafka.tools;
 
+import java.io.IOException;
```
Comment (dguy) [inline core/src/main/scala/kafka/tools/StreamsResetter.java]: Why the re-ordering of imports?

File: core/src/main/scala/kafka/tools/StreamsResetter.java [hunk 21]
```diff
@@ -219,25 +242,83 @@ private void resetInputAndInternalAndSeekToEndIntermediateTopicOffsets() {
                 }
             }
 
-            if (inputAndInternalTopicPartitions.size() > 0) {
-                client.seekToBeginning(inputAndInternalTopicPartitions);
+            maybeSeekToBeginning(client, inputAndInternalTopicPartitions, internalTopics);
+
+            maybeSeekToEnd(client, intermediateTopicPartitions);
+
+            if (!dryRun) {
```
Comment (dguy) [inline core/src/main/scala/kafka/tools/StreamsResetter.java:253]: could we move this `if(!dryRun)` to before `maybeSeekToBeginning`? Then we don't need the additional checks in `maybeSeekToBeginning` and `maybeSeekToEnd` you could just log the messages you want to log here or in another method.

**Review by dguy — COMMENTED** [review]
Thanks @bharatviswa504 for the patch.
Left a couple of comments, but i'm wondering if there is a way we can do this without having to have `if(!dryRun)` in multiple places? Can we not just have:
```
if (dryRun) {
// print everything that would happend
} else {
  // run the reset
}
```

**Comment by mjsax** [comment]
@dguy I don't see a good way to do this without the risk that the "print path" shows something else as the actual "execute path" does. We should ensure that we cannot introduce bug (both paths could diverge if we split them). Does this make sense?

File: core/src/main/scala/kafka/tools/StreamsResetter.java (hunk 20 above)
Comment (bharatviswa504) [reply core/src/main/scala/kafka/tools/StreamsResetter.java]: I have done re-ordering to follow java standard imports first and then later project specific imports.
Please let me know if you want me to revert it back

File: core/src/main/scala/kafka/tools/StreamsResetter.java (hunk 21 above)
Comment (bharatviswa504) [reply core/src/main/scala/kafka/tools/StreamsResetter.java:253]: @dguy as to have a single code path for main logic and dry-run logic it is done this way, 
The intial PR is done like a single check 
if dry-run 
{
}
else {
}

for more info can have a look into @mjsax comments. As the logic between dry-run and actual logic may change if any future changes will be done to this tool, which may show different behavior. So felt it is valid and followed this approach, even if the code does not look cleaner.

File: core/src/main/scala/kafka/tools/StreamsResetter.java (hunk 20 above)
Comment (dguy) [reply core/src/main/scala/kafka/tools/StreamsResetter.java]: We typically don't re-order imports

**Comment by dguy** [comment]
@mjsax fair enough.

**Comment by enothereska** [comment]
LGTM

File: core/src/main/scala/kafka/tools/StreamsResetter.java (hunk 20 above)
Comment (bharatviswa504) [reply core/src/main/scala/kafka/tools/StreamsResetter.java]: reverted back the order to original.

**Comment by bharatviswa504** [comment]
@mjsax @dguy addressed all the review comments.

**Comment by bharatviswa504** [comment]
Test failures are not related to this patch.

**Review by mjsax — APPROVED** [review]
LGTM. Thanks a lot for this patch @bharatviswa504 !

Call for review and merging @guozhangwang

**Comment by mjsax** [comment]
Retest this please.

**Comment by bharatviswa504** [comment]
Retest this please.

**Comment by bharatviswa504** [comment]
@guozhangwang could you please help this in reviewing and pushing changes.
As there are there other 2 jira's for streams reset tool, will create PR's once these changes are in.

**Comment by guozhangwang** [comment]
Merged to trunk.
